deployment/configenv/xml_jlibpt/Software.cpp

```diff
@@ -0,0 +1,295 @@
+
+    HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "Software.hpp"
+#include "SWProcess.hpp"
+#include "SWDropZone.hpp"
+#include "SWThorCluster.hpp"
+#include "SWRoxieCluster.hpp"
+#include "SWEspProcess.hpp"
+#include "SWEspService.hpp"
+#include "SWDirectories.hpp"
+#include "SWDaliProcess.hpp"
+#include "SWTopology.hpp"
+#include "deployutils.hpp"
+#include "ComponentBase.hpp"
+
+namespace ech
+{
+
+Software::Software(EnvHelper * envHelper):ComponentBase("software", envHelper)
+{
+}
+
+Software::~Software()
+{
+
+  HashIterator swCompIter(swCompMap);
+  ForEach(swCompIter)
+  {
+    IMapping &cur = swCompIter.query();
+    IConfigComp* pSWComp = swCompMap.mapToValue(&cur);
+    SWComponentBase *swcb = (SWComponentBase*) pSWComp;
+    //swcb-Release();
+    ::Release(swcb);
+  }
+}
+
+void Software::create(IPropertyTree *params)
+{
+  IPropertyTree * envTree = envHelper->getEnvTree();
+
+  const IPropertyTree * buildSetTree = envHelper->getBuildSetTree();
+  envTree->addPropTree(XML_TAG_SOFTWARE, createPTreeFromIPT(
+    buildSetTree->queryPropTree("./" XML_TAG_SOFTWARE)));
+
+  getSWComp("esp")->create(params);
+
+  StringBuffer xpath;
+  xpath.clear().appendf("./%s/%s/%s", XML_TAG_PROGRAMS, XML_TAG_BUILD, XML_TAG_BUILDSET);
+    Owned<IPropertyTreeIterator> buildSetInsts = buildSetTree->getElements(xpath.str());
+
+  const GenEnvRules& rules = envHelper->getGenEnvRules();
+
+  ForEach(*buildSetInsts)
+  {
+    IPropertyTree* pBuildSet = &buildSetInsts->query();
+    const char* buildSetName = pBuildSet->queryProp(XML_ATTR_NAME);
+    if (stricmp(buildSetName, "esp") == 0) continue;
+    if (rules.foundInProp("do_not_generate", buildSetName)) continue;
+
+    getSWComp(buildSetName)->create(params);
+  }
+
+}
+
+unsigned Software::add(IPropertyTree *params)
+{
+  const char *comp    = params->queryProp("@component");
+  return getSWComp(comp)->add(params);
+}
+
+void Software::modify(IPropertyTree *params)
+{
+   //IPropertyTree * envTree = envHelper->getEnvTree();
+   const char *comp = params->queryProp("@component");
+   getSWComp(comp)->modify(params);
+}
+
+
+void Software::remove(IPropertyTree *params)
+{
+   const char *comp = params->queryProp("@component");
+   return getSWComp(comp)->remove(params);
+}
```

Comment:
return type is void, looks like you are returning the value from the remove call. Should that call change its return value this will break.